src/text-pango.c

```diff
@@ -451,7 +453,7 @@ gdip_pango_setup_layout (GpGraphics *graphics, GDIPCONST WCHAR *stringUnicode, i
 			pango_layout_iter_get_line_yrange (iter, &y0, &y1);
 			//g_warning("yrange: %d  %d  clipy2: %f", y0 / PANGO_SCALE, y1 / PANGO_SCALE, clipy2);
 			/* StringFormatFlagsLineLimit */
-			if (((fmt->formatFlags & StringFormatFlagsLineLimit) && y1 / PANGO_SCALE >= clipy2) || (y0 / PANGO_SCALE >= clipy2)) {
+			if (((fmt->formatFlags & StringFormatFlagsLineLimit) && y1 / PANGO_SCALE > FrameHeight) || (y0 / PANGO_SCALE >= clipy2)) {
```

Comment:
This won't work properly when using the DrawString overloads that take point-only coordinates. After all that's why `clipy2` is there in the first place.

Reply:
This is only hit when `FrameHeight` is greater than 0 (line 437 / 439), so it shouldn't be a problem. `clipy2` only differs from `FrameHeight` if NoClip is not specified, and in that case is set based on the clip region. However, that specific change relates only to LineLimit, and LineLimit should not clip on the clip region, only on the bounds. Probably something else I should write a unit test for.

Reply:
Eh, sorry, I missed that. It was me who originally changed it from > to >=. There are some unit tests to cover the general line measurements (https://github.com/mono/libgdiplus/commit/0a27652a0ed9e87d3153b6e03e784d9868a12e1e). If I missed something then a unit test would certainly be welcome. I'm generally trying to get a good unit test coverage for the Pango code to have a more convincing case for making it the default (#269).

Reply:
I guess my intention was that the pixels at FrameHeight are already not part of the rectangle (the general formula that FrameX + FrameHeight - 1 is the last included pixel) and thus y1 >= FrameHeight is already out of the rectangle.

Reply:
True, however from looking at the Pango source, [`pango_layout_iter_get_line_yrange`](https://github.com/GNOME/pango/blob/cdb88930da73bdbeb0e6167825121435d7e5e99f/pango/pango-layout.c#L6393) is effectively [`pango_layout_iter_get_line_extents`](https://github.com/GNOME/pango/blob/cdb88930da73bdbeb0e6167825121435d7e5e99f/pango/pango-layout.c#L6352)`(iter, NULL, logical_rect); ... *y1 = logical_rect->y + logical_rect->height;` Therefore `y1` is actually not included in the line either, and making > correct. I realise the Pango documentation does not make this clear.

Reply:
I've just added a test-case for LineLimit (709a5c1).

Reply:
LGTM

Reply:
Any reason why not keep the comparison to clipy2 (instead of FrameHeight)? If it was intentional then we could drop the whole `clipy2` code, but that would certainly need a test.

Reply:
Because if NoClip is not set, clipy2 is based on the clip region. The clip region does not affect LineLimit on Windows – it is possible to have LineLimit set, and a clip region that shows a partial line, as long as the bounding box is large enough to fit the whole line.

clipy2 is used later in the line to eliminate drawing lines that are not visible, irrespective of LineLimit.

Reply:
That still sounds wrong. Imagine this example:
- The clip region would cut of last two lines
- The bounding box has space for both of the lines
You say that StringFormatFlagsLineLimit doesn't take the clipping region into account, but this loop would still break on the second condition and report wrong results.

Reply:
Since I was trying to get LineLimit working correctly I never really thought about the second condition except to figure out roughly what it was meant to do. So I never thought about the implications of it (i.e. altering bounds, number of lines / characters fitted). But you're absolutely right. That too should use FrameHeight, and we could indeed get rid of the clipy2 code.